File: src/libnxjava/java/src/main/java/org/netxms/bridge/Platform.java
/**
 * NetXMS Java Bridge
 * Copyright (C) 2013 TEMPEST a.s.
 * Copyright (C) 2014-2017 Raden Solutions
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package org.netxms.bridge;

/**
 * Platform utility class
 */
public final class Platform
{
   /**
    * Get NetXMS directory (internal bridge to native code)
    * 
    * @param type directory type
    * @return path to directory or empty string on error
    */
   protected static native String getNetXMSDirectoryInternal(int type);

   /**
    * Write log message using NetXMS logging facility
    * 
    * @param level log level
    * @param message message
    */
   protected static native void writeLog(int level, String message);

   /**
    * Write debug log message using NetXMS loggin facility
    * 
    * @param level debug level (0-9)
    * @param message message
    */
   public static native void writeDebugLog(int level, String message);

   /**
    * Wrapper for native writeLog call
    * 
    * @param level log level
    * @param message message text
    */
   public static void writeLog(LogLevel level, String message)
   {
      writeLog(level.getValue(), message);
   }

   /**
    * Write exception's stack trace to debug log
    * 
    * @param level log level
    * @param prefix message prefix
    * @param e exception to log
    */
   public static void writeDebugLog(int level, String prefix, Throwable e)
   {
      for(StackTraceElement s : e.getStackTrace())
      {
         writeDebugLog(level, prefix + s.toString());
      }
      if (e.getCause() != null)
      {
         writeDebugLog(level, prefix.trim() + " Caused by: " + e.getCause().getClass().getCanonicalName() + ": " + e.getCause().getMessage());
         writeDebugLog(level, prefix, e.getCause());
      }
   }
   
   /**
    * Get NetXMS directory
    * 
    * @param type directory type
    * @return path to directory or empty string on error
    */
   public static String getNetXMSDirectory(DirectoryType type)
   {
      return getNetXMSDirectoryInternal(type.getValue());
   }
}

File: src/libnxjava/platform.cpp
/*
** NetXMS - Network Management System
** Copyright (C) 2003-2023 Victor Kirhenshtein
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU Lesser General Public License as published
** by the Free Software Foundation; either version 3 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU Lesser General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
**
** File: platform.cpp
**/

#include "libnxjava.h"

/**
 * Class:     org.netxms.bridge.Platform
 * Method:    getNetXMSDirectoryInternal
 * Signature: (I)Ljava/lang/String;
 */
static jstring JNICALL J_getNetXMSDirectoryInternal(JNIEnv *jenv, jclass jcls, jint type)
{
   TCHAR buffer[MAX_PATH];
   GetNetXMSDirectory(static_cast<nxDirectoryType>(type), buffer);
   return JavaStringFromCString(jenv, buffer);
}

/**
 * Class:     org.netxms.bridge.Platform
 * Method:    writeLog
 * Signature: (Ljava/lang/String;ILjava/lang/String;)V
 */
static void JNICALL J_writeLog(JNIEnv *jenv, jclass jcls, jstring jtag, jint level, jstring jmessage)
{
   if (jmessage == nullptr)
      return;

   TCHAR *message = CStringFromJavaString(jenv, jmessage);
   if (jtag != nullptr)
   {
      TCHAR *tag = CStringFromJavaString(jenv, jtag);
      nxlog_write_tag(static_cast<int>(level), tag, _T("%s"), message);
      MemFree(tag);
   }
   else
   {
      nxlog_write_tag(static_cast<int>(level), _T("java"), _T("%s"), message);
   }
   MemFree(message);
}

/**
 * Class:     org.netxms.bridge.Platform
 * Method:    writeDebugLog
 * Signature: (Ljava/lang/String;ILjava/lang/String;)V
 */
static void JNICALL J_writeDebugLog(JNIEnv *jenv, jclass jcls, jstring jtag, jint level, jstring jmessage)
{
   if (jmessage == nullptr)
      return;

   TCHAR *message = CStringFromJavaString(jenv, jmessage);
   if (jtag != nullptr)
   {
      TCHAR *tag = CStringFromJavaString(jenv, jtag);
      nxlog_debug_tag(tag, static_cast<int>(level), _T("%s"), message);
      MemFree(tag);
   }
   else
   {
      nxlog_debug_tag(_T("java"), static_cast<int>(level), _T("%s"), message);
   }
   MemFree(message);
}

/**
 * Native methods
 */
static JNINativeMethod s_jniNativeMethods[] =
{
   { (char *)"getNetXMSDirectoryInternal", (char *)"(I)Ljava/lang/String;", (void *)J_getNetXMSDirectoryInternal },
   { (char *)"writeDebugLog", (char *)"(Ljava/lang/String;ILjava/lang/String;)V", (void *)J_writeDebugLog },
   { (char *)"writeLog", (char *)"(Ljava/lang/String;ILjava/lang/String;)V", (void *)J_writeLog }
};

/**
 * Register native methods for Platform classes
 */
bool RegisterPlatformNatives(JNIEnv *env)
{
   jclass platformClass = CreateJavaClassGlobalRef(env, "org/netxms/bridge/Platform");
   if (platformClass == nullptr)
      return false;

   if (env->RegisterNatives(platformClass, s_jniNativeMethods, (jint)(sizeof(s_jniNativeMethods) / sizeof (s_jniNativeMethods[0]))) != 0)
   {
      nxlog_write_tag(NXLOG_ERROR, DEBUG_TAG_JAVA_BRIDGE, _T("Failed to register native methods for platform class"));
      return false;
   }

   return true;
}
